from typing import Protocol
from ..Common._exceptions import UTMApplicationError


class IDepartmentRepository(Protocol):

    def set_resource_exception(exc: UTMApplicationError): ...

    def get_admin_by_department_id(department_id: str): ...

    def get_tutors_by_department_id(department_id: str): ...

    def get_students_by_department_id(department_id: str, filters: dict[str, str]): ...

    def get_student_by_id(student_id: str): ...

    def get_tutor_by_id(tutor_id: str): ...

    def get_personal_tutor_by_student_id(student_id: str) -> dict[str, str]: ...

    def get_personal_tutors_by_department_id(department_id: str): ...

    def get_personal_tutor_by_id(tutor_id: str): ...

    def create_personal_tutor(tutor: dict[str, str]): ...

    def update_personal_tutor(tutor: dict[str, str]): ...

    def add_cohort(cohort: dict[str, str]): ...

    def get_cohort(cohort_id: str): ...

    def get_cohorts(department_id: str): ...

    def set_cohort_tutor(
        _id: str, department: str, cohort_tutor_data: dict[str, str]
    ): ...

    def add_students_to_cohort(
        _id: str, department: str, cohort_students_data: dict[str, str]
    ): ...


class ISchoolLmsRepository(Protocol):

    def get_academic_session(academic_session_id: str): ...

    def get_academic_sessions_by_ids(academic_session_ids: list[str]): ...
